{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "check ffmode else weird syntaxis in the entire text. Example: Acta BotÄ±fmmode\\acutea\\elseÃ¡\\finica VenezuÄ±fmmode\\acutee\\elseÃ©\\filica  \n",
    "check duplicates and remove them  \n",
    "\n",
    "\n",
    "Ä±fmmode\\\\(\\w+)\\\\else.*\\\\fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re \n",
    "import numpy as np\n",
    "import os.path\n",
    "\n",
    "import bibtexparser\n",
    "from bibtexparser.bparser import BibTexParser\n",
    "from bibtexparser.customization import homogenize_latex_encoding\n",
    "from bibtexparser.customization import convert_to_unicode\n",
    "\n",
    "#define parser  method for bibtex\n",
    "bib_parser = BibTexParser()\n",
    "bib_parser.customization = convert_to_unicode\n",
    "# bib_parser.customization = homogenize_latex_encoding\n",
    "\n",
    "from fuzzywuzzy import fuzz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load abstracts databases (nucleotides and SRA)\n",
    "df_markers_raw = pd.read_excel(\"../4_Abstract_extraction_Manual/final_database.xlsx\")\n",
    "df_sra_raw = pd.read_excel(\"../4_Abstract_extraction_Manual/final_database_SRA.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_raw = pd.concat([df_markers_raw, df_sra_raw])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_Index</th>\n",
       "      <th>Old_title</th>\n",
       "      <th>Citation</th>\n",
       "      <th>Google link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SRA-3906</td>\n",
       "      <td>We included 85 samples of species of Bursera (...</td>\n",
       "      <td>No existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=We+inclu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SRA-3923</td>\n",
       "      <td>Whole-Genome Duplication and Acceleration of M...</td>\n",
       "      <td>@article{Yu2017Feb,\\n        author = {Yu, Yan...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=Whole-Ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>SRA-3931</td>\n",
       "      <td>Widespread adaptive evolution during repeated ...</td>\n",
       "      <td>@article{Nevado2016Aug,\\n        author = {Nev...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=Widespre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>SRA-3992</td>\n",
       "      <td>double digest RAD data from Berberis spp. (Ber...</td>\n",
       "      <td>No existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=double+d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>SRA-4011</td>\n",
       "      <td>phylogenetic study of Juglans</td>\n",
       "      <td>No existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=phylogen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N_Index                                          Old_title  \\\n",
       "148  SRA-3906  We included 85 samples of species of Bursera (...   \n",
       "149  SRA-3923  Whole-Genome Duplication and Acceleration of M...   \n",
       "150  SRA-3931  Widespread adaptive evolution during repeated ...   \n",
       "151  SRA-3992  double digest RAD data from Berberis spp. (Ber...   \n",
       "152  SRA-4011                      phylogenetic study of Juglans   \n",
       "\n",
       "                                              Citation  \\\n",
       "148                                          No existe   \n",
       "149  @article{Yu2017Feb,\\n        author = {Yu, Yan...   \n",
       "150  @article{Nevado2016Aug,\\n        author = {Nev...   \n",
       "151                                          No existe   \n",
       "152                                          No existe   \n",
       "\n",
       "                                           Google link  \n",
       "148  https://scholar.google.com/scholar?&q=We+inclu...  \n",
       "149  https://scholar.google.com/scholar?&q=Whole-Ge...  \n",
       "150  https://scholar.google.com/scholar?&q=Widespre...  \n",
       "151  https://scholar.google.com/scholar?&q=double+d...  \n",
       "152  https://scholar.google.com/scholar?&q=phylogen...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_raw.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy original dataframe (only first three columns)\n",
    "df = df_raw.iloc[:, 0:3].copy()\n",
    "\n",
    "#remove rows that has no bibtex citation\n",
    "df = df[df[\"Citation\"].str.startswith(\"@\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#create final df, in which tematic columns will be added.\n",
    "#in this dataframe will be the information for every citation field\n",
    "\n",
    "final_df = pd.DataFrame()\n",
    "\n",
    "\n",
    "for row in df.itertuples():\n",
    "    new_row_values = {}\n",
    "# for row in df.head(5).itertuples():\n",
    "    #parse bibtex\n",
    "    citation = bibtexparser.loads(row.Citation, bib_parser)\n",
    "    citation = citation.entries[-1]\n",
    "    \n",
    "    #add N_Index \n",
    "    # row_values[\"N_Index\"] = row.N_Index\n",
    "\n",
    "    #add columns for each citation field (added prefix c. = citation)\n",
    "    for field in citation:\n",
    "        # #verify weird syntax for special chars \n",
    "        # citation[field] = re.sub(r\"Ä±fmmode\\\\(\\w+)\\\\else.*\\\\fi\", r\"\\1\", citation[field])\n",
    "\n",
    "\n",
    "        new_row_values[f\"c.{field.lower()}\"] = citation[field]\n",
    "\n",
    "    #count names and add them to the final df\n",
    "    ##split names by \" and \"\n",
    "    try:\n",
    "        names = citation[\"author\"].split(\" and \")\n",
    "        authors = bibtexparser.customization.getnames(names)\n",
    "        n_authors = len(authors)\n",
    "    except:\n",
    "        authors = \"ERROR\"\n",
    "        n_authors = \"ERROR\"\n",
    "    \n",
    "    #replace names with a organized structure from bibtexparser\n",
    "    new_row_values[\"c.author\"] = authors\n",
    "    new_row_values[\"s.n_authors\"] = str(n_authors) #in this prefix s. is added to refer statistics\n",
    "\n",
    "    #add old title\n",
    "    new_row_values[\"a.oldtitle\"] = row.Old_title #prefix a. is for archive\n",
    "    \n",
    "    #compare old and new title\n",
    "    try:\n",
    "        titles_similitud = fuzz.token_sort_ratio(row.Old_title.lower(), citation[\"title\"].lower())\n",
    "        new_row_values[\"v.simil_titles\"] = titles_similitud #prefix v. is for verification\n",
    "    except:\n",
    "        new_row_values[\"v.simil_titles\"] = 0\n",
    "\n",
    "    row_to_add = pd.Series(new_row_values, name=row.N_Index)\n",
    "    final_df = final_df.append(row_to_add, sort=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a.oldtitle</th>\n",
       "      <th>c.abstract</th>\n",
       "      <th>c.address</th>\n",
       "      <th>c.author</th>\n",
       "      <th>c.doi</th>\n",
       "      <th>c.entrytype</th>\n",
       "      <th>c.eprint</th>\n",
       "      <th>c.id</th>\n",
       "      <th>c.issn</th>\n",
       "      <th>c.journal</th>\n",
       "      <th>...</th>\n",
       "      <th>c.note</th>\n",
       "      <th>c.number</th>\n",
       "      <th>c.pages</th>\n",
       "      <th>c.publisher</th>\n",
       "      <th>c.title</th>\n",
       "      <th>c.url</th>\n",
       "      <th>c.volume</th>\n",
       "      <th>c.year</th>\n",
       "      <th>s.n_authors</th>\n",
       "      <th>v.simil_titles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Andean-centred' genera in the short-branch cla...</td>\n",
       "      <td>Aim We test biogeographical hypotheses regardi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Pirie, Michael D., Chatrou, Lars W., Mols, Jo...</td>\n",
       "      <td>10.1111/j.1365-2699.2005.01388.x</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pirie2006Jan</td>\n",
       "      <td>1365-2699</td>\n",
       "      <td>Journal of Biogeography</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>31--46</td>\n",
       "      <td>John Wiley &amp; Sons, Ltd</td>\n",
       "      <td>`Andean-centred′ genera in the short-branch cl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>2006</td>\n",
       "      <td>5</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A 10-gene phylogeny of Solanum section Herpyst...</td>\n",
       "      <td>Premise of the study: Solanum section Herpysti...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Tepe, Eric J., Farruggia, Frank T., Bohs, Lynn]</td>\n",
       "      <td>10.3732/ajb.1000516</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tepe2011Aug</td>\n",
       "      <td>1537-2197</td>\n",
       "      <td>American Journal of Botany</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>1356--1365</td>\n",
       "      <td>John Wiley &amp; Sons, Ltd</td>\n",
       "      <td>A 10-gene phylogeny of Solanum section Herpyst...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98</td>\n",
       "      <td>2011</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A 250 plastome phylogeny of the grass family (...</td>\n",
       "      <td>The systematics of grasses has advanced throug...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Saarela, Jeffery M., Burke, Sean V., Wysocki,...</td>\n",
       "      <td>10.7717/peerj.4299</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saarela2018Feb</td>\n",
       "      <td>2167-8359</td>\n",
       "      <td>PeerJ</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e4299</td>\n",
       "      <td>PeerJ Inc.</td>\n",
       "      <td>A 250 plastome phylogeny of the grass family (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>10</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A 313 plastome phylogenomic analysis of Pooide...</td>\n",
       "      <td>In this study, we analyzed 313 plastid genomes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Orton, Lauren M., Barberıfmmode\\acutea\\elseá\\...</td>\n",
       "      <td>10.1016/j.ympev.2021.107110</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Orton2021Jun</td>\n",
       "      <td>1055-7903</td>\n",
       "      <td>Molecular Phylogenetics and Evolution</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107110</td>\n",
       "      <td>Academic Press</td>\n",
       "      <td>A 313 plastome phylogenomic analysis of Pooide...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>159</td>\n",
       "      <td>2021</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A Chloroplast Phylogeny of Agavaceae subfamily...</td>\n",
       "      <td>Abstractıfmmode---\\else—\\fi Agavaceae subfamil...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Halpin, Kate M., Fishbein, Mark]</td>\n",
       "      <td>10.1600/036364413X674850</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halpin2013Dec</td>\n",
       "      <td>0363-6445</td>\n",
       "      <td>Systematic Botany</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>996--1011</td>\n",
       "      <td>American Society of Plant Taxonomists</td>\n",
       "      <td>A Chloroplast Phylogeny of Agavaceae subfamily...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38</td>\n",
       "      <td>2013</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA-3855</th>\n",
       "      <td>Using transcriptomes to explore the evolution ...</td>\n",
       "      <td>Caryophyllales are a highly diverse and large ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Pucker, Boas, Feng, Tao, Brockington, Samuel F.]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>article</td>\n",
       "      <td>646133</td>\n",
       "      <td>Pucker2020Feb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bioRxiv</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>646133</td>\n",
       "      <td>Cold Spring Harbor Laboratory</td>\n",
       "      <td>Next generation sequencing to investigate geno...</td>\n",
       "      <td>https://doi.org/10.1101/646133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020</td>\n",
       "      <td>3</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA-3870</th>\n",
       "      <td>Viburnum lentago clade RADseq data</td>\n",
       "      <td>Premise We take an integrative approach in ass...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Spriggs, Elizabeth L., Schlutius, Caroline, E...</td>\n",
       "      <td>10.1002/ajb2.1292</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spriggs2019Jun</td>\n",
       "      <td>0002-9122</td>\n",
       "      <td>American Journal of Botany</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>833--849</td>\n",
       "      <td>John Wiley &amp; Sons, Ltd</td>\n",
       "      <td>Differences in flowering time maintain species...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106</td>\n",
       "      <td>2019</td>\n",
       "      <td>7</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA-3871</th>\n",
       "      <td>Viburnum nudum complex RADseq data</td>\n",
       "      <td>Species are the starting point for most studie...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Spriggs, Elizabeth L., Eaton, Deren A. R., Sw...</td>\n",
       "      <td>10.1093/sysbio/syy084</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spriggs2019Mar</td>\n",
       "      <td>1063-5157</td>\n",
       "      <td>Systematic Biology</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>187--203</td>\n",
       "      <td>Oxford Academic</td>\n",
       "      <td>Restriction-Site-Associated DNA Sequencing Rev...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "      <td>2019</td>\n",
       "      <td>6</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA-3923</th>\n",
       "      <td>Whole-Genome Duplication and Acceleration of M...</td>\n",
       "      <td>The pattern and rate of genome evolution have ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Yu, Yan, Xiang, Qiuyun, Manos, Paul S., Solti...</td>\n",
       "      <td>10.1371/journal.pone.0171361</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yu2017Feb</td>\n",
       "      <td>1932-6203</td>\n",
       "      <td>PLOS ONE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>e0171361</td>\n",
       "      <td>Public Library of Science</td>\n",
       "      <td>Whole-genome duplication and molecular evoluti...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>2017</td>\n",
       "      <td>9</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRA-3931</th>\n",
       "      <td>Widespread adaptive evolution during repeated ...</td>\n",
       "      <td>The evolutionary processes that drive rapid sp...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Nevado, Bruno, Atchison, Guy W., Hughes, Coli...</td>\n",
       "      <td>10.1038/ncomms12384</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nevado2016Aug</td>\n",
       "      <td>2041-1723</td>\n",
       "      <td>Nature Communications</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12384</td>\n",
       "      <td>1--9</td>\n",
       "      <td>Nature Publishing Group</td>\n",
       "      <td>Widespread adaptive evolution during repeated ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>2016</td>\n",
       "      <td>4</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3948 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 a.oldtitle  \\\n",
       "0         Andean-centred' genera in the short-branch cla...   \n",
       "23        A 10-gene phylogeny of Solanum section Herpyst...   \n",
       "26        A 250 plastome phylogeny of the grass family (...   \n",
       "29        A 313 plastome phylogenomic analysis of Pooide...   \n",
       "32        A Chloroplast Phylogeny of Agavaceae subfamily...   \n",
       "...                                                     ...   \n",
       "SRA-3855  Using transcriptomes to explore the evolution ...   \n",
       "SRA-3870                 Viburnum lentago clade RADseq data   \n",
       "SRA-3871                 Viburnum nudum complex RADseq data   \n",
       "SRA-3923  Whole-Genome Duplication and Acceleration of M...   \n",
       "SRA-3931  Widespread adaptive evolution during repeated ...   \n",
       "\n",
       "                                                 c.abstract c.address  \\\n",
       "0         Aim We test biogeographical hypotheses regardi...       NaN   \n",
       "23        Premise of the study: Solanum section Herpysti...       NaN   \n",
       "26        The systematics of grasses has advanced throug...       NaN   \n",
       "29        In this study, we analyzed 313 plastid genomes...       NaN   \n",
       "32        Abstractıfmmode---\\else—\\fi Agavaceae subfamil...       NaN   \n",
       "...                                                     ...       ...   \n",
       "SRA-3855  Caryophyllales are a highly diverse and large ...       NaN   \n",
       "SRA-3870  Premise We take an integrative approach in ass...       NaN   \n",
       "SRA-3871  Species are the starting point for most studie...       NaN   \n",
       "SRA-3923  The pattern and rate of genome evolution have ...       NaN   \n",
       "SRA-3931  The evolutionary processes that drive rapid sp...       NaN   \n",
       "\n",
       "                                                   c.author  \\\n",
       "0         [Pirie, Michael D., Chatrou, Lars W., Mols, Jo...   \n",
       "23         [Tepe, Eric J., Farruggia, Frank T., Bohs, Lynn]   \n",
       "26        [Saarela, Jeffery M., Burke, Sean V., Wysocki,...   \n",
       "29        [Orton, Lauren M., Barberıfmmode\\acutea\\elseá\\...   \n",
       "32                        [Halpin, Kate M., Fishbein, Mark]   \n",
       "...                                                     ...   \n",
       "SRA-3855  [Pucker, Boas, Feng, Tao, Brockington, Samuel F.]   \n",
       "SRA-3870  [Spriggs, Elizabeth L., Schlutius, Caroline, E...   \n",
       "SRA-3871  [Spriggs, Elizabeth L., Eaton, Deren A. R., Sw...   \n",
       "SRA-3923  [Yu, Yan, Xiang, Qiuyun, Manos, Paul S., Solti...   \n",
       "SRA-3931  [Nevado, Bruno, Atchison, Guy W., Hughes, Coli...   \n",
       "\n",
       "                                     c.doi c.entrytype c.eprint  \\\n",
       "0         10.1111/j.1365-2699.2005.01388.x     article      NaN   \n",
       "23                     10.3732/ajb.1000516     article      NaN   \n",
       "26                      10.7717/peerj.4299     article      NaN   \n",
       "29             10.1016/j.ympev.2021.107110     article      NaN   \n",
       "32                10.1600/036364413X674850     article      NaN   \n",
       "...                                    ...         ...      ...   \n",
       "SRA-3855                               NaN     article   646133   \n",
       "SRA-3870                 10.1002/ajb2.1292     article      NaN   \n",
       "SRA-3871             10.1093/sysbio/syy084     article      NaN   \n",
       "SRA-3923      10.1371/journal.pone.0171361     article      NaN   \n",
       "SRA-3931               10.1038/ncomms12384     article      NaN   \n",
       "\n",
       "                    c.id     c.issn                              c.journal  \\\n",
       "0           Pirie2006Jan  1365-2699                Journal of Biogeography   \n",
       "23           Tepe2011Aug  1537-2197             American Journal of Botany   \n",
       "26        Saarela2018Feb  2167-8359                                  PeerJ   \n",
       "29          Orton2021Jun  1055-7903  Molecular Phylogenetics and Evolution   \n",
       "32         Halpin2013Dec  0363-6445                      Systematic Botany   \n",
       "...                  ...        ...                                    ...   \n",
       "SRA-3855   Pucker2020Feb        NaN                                bioRxiv   \n",
       "SRA-3870  Spriggs2019Jun  0002-9122             American Journal of Botany   \n",
       "SRA-3871  Spriggs2019Mar  1063-5157                     Systematic Biology   \n",
       "SRA-3923       Yu2017Feb  1932-6203                               PLOS ONE   \n",
       "SRA-3931   Nevado2016Aug  2041-1723                  Nature Communications   \n",
       "\n",
       "          ... c.note c.number     c.pages  \\\n",
       "0         ...    NaN        1      31--46   \n",
       "23        ...    NaN        8  1356--1365   \n",
       "26        ...    NaN      NaN       e4299   \n",
       "29        ...    NaN      NaN      107110   \n",
       "32        ...    NaN        4   996--1011   \n",
       "...       ...    ...      ...         ...   \n",
       "SRA-3855  ...    NaN      NaN      646133   \n",
       "SRA-3870  ...    NaN        6    833--849   \n",
       "SRA-3871  ...    NaN        2    187--203   \n",
       "SRA-3923  ...    NaN        2    e0171361   \n",
       "SRA-3931  ...    NaN    12384        1--9   \n",
       "\n",
       "                                    c.publisher  \\\n",
       "0                        John Wiley & Sons, Ltd   \n",
       "23                       John Wiley & Sons, Ltd   \n",
       "26                                   PeerJ Inc.   \n",
       "29                               Academic Press   \n",
       "32        American Society of Plant Taxonomists   \n",
       "...                                         ...   \n",
       "SRA-3855          Cold Spring Harbor Laboratory   \n",
       "SRA-3870                 John Wiley & Sons, Ltd   \n",
       "SRA-3871                        Oxford Academic   \n",
       "SRA-3923              Public Library of Science   \n",
       "SRA-3931                Nature Publishing Group   \n",
       "\n",
       "                                                    c.title  \\\n",
       "0         `Andean-centred′ genera in the short-branch cl...   \n",
       "23        A 10-gene phylogeny of Solanum section Herpyst...   \n",
       "26        A 250 plastome phylogeny of the grass family (...   \n",
       "29        A 313 plastome phylogenomic analysis of Pooide...   \n",
       "32        A Chloroplast Phylogeny of Agavaceae subfamily...   \n",
       "...                                                     ...   \n",
       "SRA-3855  Next generation sequencing to investigate geno...   \n",
       "SRA-3870  Differences in flowering time maintain species...   \n",
       "SRA-3871  Restriction-Site-Associated DNA Sequencing Rev...   \n",
       "SRA-3923  Whole-genome duplication and molecular evoluti...   \n",
       "SRA-3931  Widespread adaptive evolution during repeated ...   \n",
       "\n",
       "                                   c.url c.volume c.year s.n_authors  \\\n",
       "0                                    NaN       33   2006           5   \n",
       "23                                   NaN       98   2011           3   \n",
       "26                                   NaN        6   2018          10   \n",
       "29                                   NaN      159   2021           7   \n",
       "32                                   NaN       38   2013           2   \n",
       "...                                  ...      ...    ...         ...   \n",
       "SRA-3855  https://doi.org/10.1101/646133      NaN   2020           3   \n",
       "SRA-3870                             NaN      106   2019           7   \n",
       "SRA-3871                             NaN       68   2019           6   \n",
       "SRA-3923                             NaN       12   2017           9   \n",
       "SRA-3931                             NaN        7   2016           4   \n",
       "\n",
       "         v.simil_titles  \n",
       "0                 100.0  \n",
       "23                100.0  \n",
       "26                100.0  \n",
       "29                100.0  \n",
       "32                100.0  \n",
       "...                 ...  \n",
       "SRA-3855           50.0  \n",
       "SRA-3870           43.0  \n",
       "SRA-3871           34.0  \n",
       "SRA-3923           71.0  \n",
       "SRA-3931           89.0  \n",
       "\n",
       "[3948 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove duplicates\n",
    "final_df = final_df.drop_duplicates(subset=\"c.title\", keep=\"first\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export database as csv\n",
    "# in this database tematic extraction will be done\n",
    "final_df.to_csv(\"database_for_tematic_extraction.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
