{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO\n",
    "~~only consider those that has @ as initial char in Citation columns~~   \n",
    "~~add one column per tematic~~  \n",
    "add column with comparison between old tile and new title  \n",
    "~~only maintain N_index, from previous df~~  \n",
    "~~check how parse bibtex, or do it manually (first prefered due to some latex in the text)~~~  \n",
    "~~develop a function to do tematics one by one, in that way I can add manually new tematics without reload everything~~   \n",
    "check ffmode else weird syntaxis in the entire text. Example: Acta BotÄ±fmmode\\acutea\\elseÃ¡\\finica VenezuÄ±fmmode\\acutee\\elseÃ©\\filica  \n",
    "check duplicates and remove them  \n",
    "\n",
    "\n",
    "Ä±fmmode\\\\(\\w+)\\\\else.*\\\\fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re \n",
    "import numpy as np\n",
    "import os.path\n",
    "\n",
    "import bibtexparser\n",
    "from bibtexparser.bparser import BibTexParser\n",
    "from bibtexparser.customization import homogenize_latex_encoding\n",
    "from bibtexparser.customization import convert_to_unicode\n",
    "\n",
    "#define parser  method for bibtex\n",
    "bib_parser = BibTexParser()\n",
    "bib_parser.customization = convert_to_unicode\n",
    "# bib_parser.customization = homogenize_latex_encoding\n",
    "\n",
    "from fuzzywuzzy import fuzz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load abstracts database\n",
    "df_raw = pd.read_csv(\"../4_Abstract_extraction_Manual/DATABASE - DATABASE_preview.csv\", low_memory = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N_Index</th>\n",
       "      <th>Old_title</th>\n",
       "      <th>Citation</th>\n",
       "      <th>Google link</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Systematic</th>\n",
       "      <th>Mexico</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>First gen.</th>\n",
       "      <th>...</th>\n",
       "      <th>Parsimony</th>\n",
       "      <th>Likelihood</th>\n",
       "      <th>Bayesian</th>\n",
       "      <th>MSC</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Notas</th>\n",
       "      <th>Autocheck: Abstract integrity (7)</th>\n",
       "      <th>Autocheck: Abbrev. (22)</th>\n",
       "      <th>Autocheck: Short abstract (22)</th>\n",
       "      <th>Carlos doublecheck (170)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Andean-centred' genera in the short-branch cla...</td>\n",
       "      <td>@article{Pirie2006Jan,\\n        author = {Piri...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=%27Andea...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Mediterranean' disjunction in Senecio sect. Se...</td>\n",
       "      <td>no existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=%27Medit...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>What are natural clades in the Caesalpinia gro...</td>\n",
       "      <td>no existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=%27What+...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>5S NTS of chili peppers (Capsicum-Solanaceae):...</td>\n",
       "      <td>no existe</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=5S+NTS+o...</td>\n",
       "      <td>Auto</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23</td>\n",
       "      <td>A 10-gene phylogeny of Solanum section Herpyst...</td>\n",
       "      <td>@article{Tepe2011Aug,\\n        author = {Tepe,...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=A+10-gen...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4986</th>\n",
       "      <td>12153</td>\n",
       "      <td>rbcL sequences provide phylogenetic insights a...</td>\n",
       "      <td>@article{Haufler1995Oct,\\n        author = {Ha...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=rbcL+seq...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4987</th>\n",
       "      <td>12154</td>\n",
       "      <td>rbcL sequences: a promising tool for the molec...</td>\n",
       "      <td>@article{Dubuisson1997Oct,\\n        author = {...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=rbcL+seq...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988</th>\n",
       "      <td>12163</td>\n",
       "      <td>trnL-F and rpl16 Sequence Data and Dense Taxon...</td>\n",
       "      <td>@article{Del-Pino2009Mar,\\n        author = {D...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=trnL-F+a...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4989</th>\n",
       "      <td>12164</td>\n",
       "      <td>ural Delineation, Molecular Phylogeny and Flor...</td>\n",
       "      <td>@article{Roquet2008Mar,\\n        author = {Roq...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=ural+Del...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4990</th>\n",
       "      <td>12165</td>\n",
       "      <td>ving relationships within Valerianaceae (Dipsa...</td>\n",
       "      <td>@article{Bell2015Feb,\\n        author = {Bell,...</td>\n",
       "      <td>https://scholar.google.com/scholar?&amp;q=ving+rel...</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Auto</td>\n",
       "      <td>...</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4991 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      N_Index                                          Old_title  \\\n",
       "0           0  Andean-centred' genera in the short-branch cla...   \n",
       "1           2  Mediterranean' disjunction in Senecio sect. Se...   \n",
       "2           4  What are natural clades in the Caesalpinia gro...   \n",
       "3          18  5S NTS of chili peppers (Capsicum-Solanaceae):...   \n",
       "4          23  A 10-gene phylogeny of Solanum section Herpyst...   \n",
       "...       ...                                                ...   \n",
       "4986    12153  rbcL sequences provide phylogenetic insights a...   \n",
       "4987    12154  rbcL sequences: a promising tool for the molec...   \n",
       "4988    12163  trnL-F and rpl16 Sequence Data and Dense Taxon...   \n",
       "4989    12164  ural Delineation, Molecular Phylogeny and Flor...   \n",
       "4990    12165  ving relationships within Valerianaceae (Dipsa...   \n",
       "\n",
       "                                               Citation  \\\n",
       "0     @article{Pirie2006Jan,\\n        author = {Piri...   \n",
       "1                                             no existe   \n",
       "2                                             no existe   \n",
       "3                                             no existe   \n",
       "4     @article{Tepe2011Aug,\\n        author = {Tepe,...   \n",
       "...                                                 ...   \n",
       "4986  @article{Haufler1995Oct,\\n        author = {Ha...   \n",
       "4987  @article{Dubuisson1997Oct,\\n        author = {...   \n",
       "4988  @article{Del-Pino2009Mar,\\n        author = {D...   \n",
       "4989  @article{Roquet2008Mar,\\n        author = {Roq...   \n",
       "4990  @article{Bell2015Feb,\\n        author = {Bell,...   \n",
       "\n",
       "                                            Google link Taxonomy Systematic  \\\n",
       "0     https://scholar.google.com/scholar?&q=%27Andea...        1          1   \n",
       "1     https://scholar.google.com/scholar?&q=%27Medit...        ?       Auto   \n",
       "2     https://scholar.google.com/scholar?&q=%27What+...        ?       Auto   \n",
       "3     https://scholar.google.com/scholar?&q=5S+NTS+o...     Auto       Auto   \n",
       "4     https://scholar.google.com/scholar?&q=A+10-gen...        0          1   \n",
       "...                                                 ...      ...        ...   \n",
       "4986  https://scholar.google.com/scholar?&q=rbcL+seq...        ?       Auto   \n",
       "4987  https://scholar.google.com/scholar?&q=rbcL+seq...        ?       Auto   \n",
       "4988  https://scholar.google.com/scholar?&q=trnL-F+a...        ?       Auto   \n",
       "4989  https://scholar.google.com/scholar?&q=ural+Del...        ?       Auto   \n",
       "4990  https://scholar.google.com/scholar?&q=ving+rel...        ?       Auto   \n",
       "\n",
       "     Mexico Unnamed: 7 Unnamed: 8 First gen.  ... Parsimony Likelihood  \\\n",
       "0         0          1          0          1  ...         1          0   \n",
       "1         ?          ?          ?          ?  ...         ?          ?   \n",
       "2         ?          ?          ?          ?  ...         ?          ?   \n",
       "3         ?          ?          ?          ?  ...         ?          ?   \n",
       "4         0          1          0          1  ...         1          0   \n",
       "...     ...        ...        ...        ...  ...       ...        ...   \n",
       "4986      ?          ?          ?       Auto  ...         ?          ?   \n",
       "4987      ?          ?          ?       Auto  ...         ?          ?   \n",
       "4988      ?          ?          ?       Auto  ...         ?          ?   \n",
       "4989      ?          ?          ?          ?  ...         ?          ?   \n",
       "4990      ?          ?          ?       Auto  ...         ?          ?   \n",
       "\n",
       "     Bayesian MSC Unnamed: 18 Notas Autocheck: Abstract integrity (7)  \\\n",
       "0           1   0         NaN   NaN                               NaN   \n",
       "1           ?   ?         NaN   NaN                               NaN   \n",
       "2           ?   ?         NaN   NaN                               NaN   \n",
       "3           ?   ?         NaN   NaN                               NaN   \n",
       "4           1   0         NaN   NaN                               NaN   \n",
       "...       ...  ..         ...   ...                               ...   \n",
       "4986        ?   ?         NaN   NaN                               NaN   \n",
       "4987        ?   ?         NaN   NaN                               NaN   \n",
       "4988        ?   ?         NaN   NaN                               NaN   \n",
       "4989        ?   ?         NaN   NaN                               NaN   \n",
       "4990        ?   ?         NaN   NaN                               NaN   \n",
       "\n",
       "     Autocheck: Abbrev. (22) Autocheck: Short abstract (22)  \\\n",
       "0                        NaN                            NaN   \n",
       "1                        NaN                            NaN   \n",
       "2                        NaN                            NaN   \n",
       "3                        NaN                            NaN   \n",
       "4                        NaN                            NaN   \n",
       "...                      ...                            ...   \n",
       "4986                     NaN                            NaN   \n",
       "4987                     NaN                            NaN   \n",
       "4988                     NaN                            NaN   \n",
       "4989                     NaN                            NaN   \n",
       "4990                     NaN                            NaN   \n",
       "\n",
       "     Carlos doublecheck (170)  \n",
       "0                         NaN  \n",
       "1                         NaN  \n",
       "2                         NaN  \n",
       "3                         NaN  \n",
       "4                         NaN  \n",
       "...                       ...  \n",
       "4986                      NaN  \n",
       "4987                      NaN  \n",
       "4988                      NaN  \n",
       "4989                      NaN  \n",
       "4990                      NaN  \n",
       "\n",
       "[4991 rows x 24 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of markers for detec sanger, nuclear, and chloroplast info\n",
    "nuclear_markers = \"Internal transcribed spacer|ITS1|ITS2|ETS|nrITS|18S|59S|5.8s|GBSSI|COSII|25S|Ypr10|pgiC|ETS|ADH2F|ADH3R|Aat|Skdh|Pgi.?2|Tpi.?1|Tpi.?2|AFLP\"\n",
    "chl_markers = \"trnL|trnG|trnR|trnT|trnS|trnF|trnH|trnQ|trnD|trnY|trnE|atpB|rubisco|rbcL|psbA|matk|psbJ|petA|ndhF|psaA|ycf3|ycf4|rpl16|trnSfM|trnfM|rps4|cemA|psbM|trnDGUC\"\n",
    "\n",
    "\n",
    "               \n",
    "# set catagories of search\n",
    "categories = {\n",
    "    \"Mexico\": \"(?<!new)(?<!nuevo)\\s+m.[xj]ic\",\n",
    "    \"Morphology\": \"morpholog|morfolog\",\n",
    "    \"First gen.\": f\"{nuclear_markers}|{chl_markers}| sanger | marker| nuclear region| chloroplast region|regi.n.*? nucr|regi.n.*? cloro| spacer|AFLP|Microsat.ll?ite\",\n",
    "    \"Second gen.\": \"radseq|angiosperm353|hybseq|454|illumina|iontorrent|whole genome|plastome\",\n",
    "    \"Third gen.\": \"pacbio|nanopore\",\n",
    "    \"Nuclear info\": f\"nuclear|{nuclear_markers}\",\n",
    "    \"Chloroplast info\": f\"c.?lorop|{chl_markers}\",\n",
    "    \"Parsimony\": \"parsimon\",\n",
    "    \"Maximum likelihood\": \" ML |maximum likelihood|verosimilitud\",\n",
    "    \"Bayesian\": \"bayes| beast\",\n",
    "    \"MSC\": \" astral | astrid | bpp |coalescent\",\n",
    "    \"New species\": \"new species|nueva. especie\",\n",
    "    \"New genus\": \"new genus|nueva.? g.nero\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mexico: None\n",
      "Morphology: None\n",
      "First gen.: <re.Match object; span=(66, 70), match='atpB'>\n",
      "Second gen.: None\n",
      "Third gen.: None\n",
      "Nuclear info: None\n",
      "Chloroplast info: <re.Match object; span=(66, 70), match='atpB'>\n",
      "Parsimony: None\n",
      "Maximum likelihood: None\n",
      "Bayesian: None\n",
      "MSC: None\n",
      "New species: None\n",
      "New genus: None\n"
     ]
    }
   ],
   "source": [
    "test = \"A Molecular Phylogenetic Study of the Palmae (Arecaceae) Based on atpB, rbcL, and  nrDNA Sequences\"\n",
    "for k in categories:\n",
    "    print(f\"{k}: {re.search(categories[k], test, re.IGNORECASE)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy original dataframe (only first three columns)\n",
    "df = df_raw.iloc[:, 0:3].copy()\n",
    "\n",
    "#remove rows that has no bibtex citation\n",
    "df = df[df[\"Citation\"].str.startswith(\"@\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#create final df, in which tematic columns will be added.\n",
    "#in this dataframe will be the information for every citation field\n",
    "\n",
    "final_df = pd.DataFrame()\n",
    "\n",
    "\n",
    "for row in df.itertuples():\n",
    "    new_row_values = {}\n",
    "# for row in df.head(5).itertuples():\n",
    "    #parse bibtex\n",
    "    citation = bibtexparser.loads(row.Citation, bib_parser)\n",
    "    citation = citation.entries[-1]\n",
    "    \n",
    "    #add N_Index \n",
    "    # row_values[\"N_Index\"] = row.N_Index\n",
    "\n",
    "    #add columns for each citation field (added prefix c. = citation)\n",
    "    for field in citation:\n",
    "        # #verify weird syntax for special chars \n",
    "        # citation[field] = re.sub(r\"Ä±fmmode\\\\(\\w+)\\\\else.*\\\\fi\", r\"\\1\", citation[field])\n",
    "\n",
    "\n",
    "        new_row_values[f\"c.{field.lower()}\"] = citation[field]\n",
    "\n",
    "    #count names and add them to the final df\n",
    "    ##split names by \" and \"\n",
    "    try:\n",
    "        names = citation[\"author\"].split(\" and \")\n",
    "        authors = bibtexparser.customization.getnames(names)\n",
    "        n_authors = len(authors)\n",
    "    except:\n",
    "        authors = \"ERROR\"\n",
    "        n_authors = \"ERROR\"\n",
    "    \n",
    "    #replace names with a organized structure from bibtexparser\n",
    "    new_row_values[\"c.author\"] = authors\n",
    "    new_row_values[\"s.n_authors\"] = str(n_authors) #in this prefix s. is added to refer statistics\n",
    "\n",
    "    #add old title\n",
    "    new_row_values[\"a.oldtitle\"] = row.Old_title #prefix a. is for archive\n",
    "    \n",
    "    #compare old and new title\n",
    "    try:\n",
    "        titles_similitud = fuzz.token_sort_ratio(row.Old_title.lower(), citation[\"title\"].lower())\n",
    "        new_row_values[\"v.simil_titles\"] = titles_similitud #prefix v. is for verification\n",
    "    except:\n",
    "        new_row_values[\"v.simil_titles\"] = 0\n",
    "\n",
    "    row_to_add = pd.Series(new_row_values, name=row.N_Index)\n",
    "    final_df = final_df.append(row_to_add, sort=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a.oldtitle</th>\n",
       "      <th>c.abstract</th>\n",
       "      <th>c.address</th>\n",
       "      <th>c.author</th>\n",
       "      <th>c.booktitle</th>\n",
       "      <th>c.doi</th>\n",
       "      <th>c.entrytype</th>\n",
       "      <th>c.eprint</th>\n",
       "      <th>c.id</th>\n",
       "      <th>c.isbn</th>\n",
       "      <th>...</th>\n",
       "      <th>c.note</th>\n",
       "      <th>c.number</th>\n",
       "      <th>c.pages</th>\n",
       "      <th>c.publisher</th>\n",
       "      <th>c.title</th>\n",
       "      <th>c.url</th>\n",
       "      <th>c.volume</th>\n",
       "      <th>c.year</th>\n",
       "      <th>s.n_authors</th>\n",
       "      <th>v.simil_titles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Andean-centred' genera in the short-branch cla...</td>\n",
       "      <td>Aim We test biogeographical hypotheses regardi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Pirie, Michael D., Chatrou, Lars W., Mols, Jo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/j.1365-2699.2005.01388.x</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pirie2006Jan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>31--46</td>\n",
       "      <td>John Wiley &amp; Sons, Ltd</td>\n",
       "      <td>`Andean-centred′ genera in the short-branch cl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>2006</td>\n",
       "      <td>5</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A 10-gene phylogeny of Solanum section Herpyst...</td>\n",
       "      <td>Premise of the study: Solanum section Herpysti...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Tepe, Eric J., Farruggia, Frank T., Bohs, Lynn]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3732/ajb.1000516</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tepe2011Aug</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>1356--1365</td>\n",
       "      <td>John Wiley &amp; Sons, Ltd</td>\n",
       "      <td>A 10-gene phylogeny of Solanum section Herpyst...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98</td>\n",
       "      <td>2011</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A 250 plastome phylogeny of the grass family (...</td>\n",
       "      <td>The systematics of grasses has advanced throug...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Saarela, Jeffery M., Burke, Sean V., Wysocki,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.7717/peerj.4299</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saarela2018Feb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>e4299</td>\n",
       "      <td>PeerJ Inc.</td>\n",
       "      <td>A 250 plastome phylogeny of the grass family (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>2018</td>\n",
       "      <td>10</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A 313 plastome phylogenomic analysis of Pooide...</td>\n",
       "      <td>In this study, we analyzed 313 plastid genomes...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Orton, Lauren M., Barberıfmmode\\acutea\\elseá\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.ympev.2021.107110</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Orton2021Jun</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107110</td>\n",
       "      <td>Academic Press</td>\n",
       "      <td>A 313 plastome phylogenomic analysis of Pooide...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>159</td>\n",
       "      <td>2021</td>\n",
       "      <td>7</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A Chloroplast Phylogeny of Agavaceae subfamily...</td>\n",
       "      <td>Abstractıfmmode---\\else—\\fi Agavaceae subfamil...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Halpin, Kate M., Fishbein, Mark]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1600/036364413X674850</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Halpin2013Dec</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>996--1011</td>\n",
       "      <td>American Society of Plant Taxonomists</td>\n",
       "      <td>A Chloroplast Phylogeny of Agavaceae subfamily...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38</td>\n",
       "      <td>2013</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12153</th>\n",
       "      <td>rbcL sequences provide phylogenetic insights a...</td>\n",
       "      <td>Opinions on the phylogeny and classification o...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Haufler, Christopher H., Ranker, Tom A.]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haufler1995Oct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>361--374</td>\n",
       "      <td>American Fern Society</td>\n",
       "      <td>RbcL Sequences Provide Phylogenetic Insights a...</td>\n",
       "      <td>http://www.jstor.org/stable/1547815</td>\n",
       "      <td>85</td>\n",
       "      <td>1995</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12154</th>\n",
       "      <td>rbcL sequences: a promising tool for the molec...</td>\n",
       "      <td>A phylogenetic study usingrbcL sequences was p...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Dubuisson, J.-Y.]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1006/mpev.1997.0414</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dubuisson1997Oct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>128--138</td>\n",
       "      <td>Academic Press</td>\n",
       "      <td>rbcL Sequences: A Promising Tool for the Molec...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12163</th>\n",
       "      <td>trnL-F and rpl16 Sequence Data and Dense Taxon...</td>\n",
       "      <td>The Amaranthaceae-Chenopodiaceae alliance has ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Del-Pino, Ivonne Sıfmmode\\acutea\\elseá\\finche...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1600/036364409787602401</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Del-Pino2009Mar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>57--67</td>\n",
       "      <td>American Society of Plant Taxonomists</td>\n",
       "      <td>trnL-F and rpl16 Sequence Data and Dense Taxon...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12164</th>\n",
       "      <td>ural Delineation, Molecular Phylogeny and Flor...</td>\n",
       "      <td>The circumscription and infrageneric classific...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Roquet, Cristina, Sıfmmode\\acutea\\elseá\\fiez,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1600/036364408783887465</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Roquet2008Mar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>203--217</td>\n",
       "      <td>American Society of Plant Taxonomists</td>\n",
       "      <td>Natural Delineation, Molecular Phylogeny and F...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>2008</td>\n",
       "      <td>6</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12165</th>\n",
       "      <td>ving relationships within Valerianaceae (Dipsa...</td>\n",
       "      <td>Abstractıfmmode---\\else—\\fi Primers from a rec...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Bell, Charles D., Calderon, Gloria, Gonzalez,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1600/036364415X686611</td>\n",
       "      <td>article</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bell2015Feb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>327--335</td>\n",
       "      <td>American Society of Plant Taxonomists</td>\n",
       "      <td>Resolving Relationships within Valerianaceae (...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40</td>\n",
       "      <td>2015</td>\n",
       "      <td>5</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3795 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              a.oldtitle  \\\n",
       "0      Andean-centred' genera in the short-branch cla...   \n",
       "23     A 10-gene phylogeny of Solanum section Herpyst...   \n",
       "26     A 250 plastome phylogeny of the grass family (...   \n",
       "29     A 313 plastome phylogenomic analysis of Pooide...   \n",
       "32     A Chloroplast Phylogeny of Agavaceae subfamily...   \n",
       "...                                                  ...   \n",
       "12153  rbcL sequences provide phylogenetic insights a...   \n",
       "12154  rbcL sequences: a promising tool for the molec...   \n",
       "12163  trnL-F and rpl16 Sequence Data and Dense Taxon...   \n",
       "12164  ural Delineation, Molecular Phylogeny and Flor...   \n",
       "12165  ving relationships within Valerianaceae (Dipsa...   \n",
       "\n",
       "                                              c.abstract c.address  \\\n",
       "0      Aim We test biogeographical hypotheses regardi...       NaN   \n",
       "23     Premise of the study: Solanum section Herpysti...       NaN   \n",
       "26     The systematics of grasses has advanced throug...       NaN   \n",
       "29     In this study, we analyzed 313 plastid genomes...       NaN   \n",
       "32     Abstractıfmmode---\\else—\\fi Agavaceae subfamil...       NaN   \n",
       "...                                                  ...       ...   \n",
       "12153  Opinions on the phylogeny and classification o...       NaN   \n",
       "12154  A phylogenetic study usingrbcL sequences was p...       NaN   \n",
       "12163  The Amaranthaceae-Chenopodiaceae alliance has ...       NaN   \n",
       "12164  The circumscription and infrageneric classific...       NaN   \n",
       "12165  Abstractıfmmode---\\else—\\fi Primers from a rec...       NaN   \n",
       "\n",
       "                                                c.author c.booktitle  \\\n",
       "0      [Pirie, Michael D., Chatrou, Lars W., Mols, Jo...         NaN   \n",
       "23      [Tepe, Eric J., Farruggia, Frank T., Bohs, Lynn]         NaN   \n",
       "26     [Saarela, Jeffery M., Burke, Sean V., Wysocki,...         NaN   \n",
       "29     [Orton, Lauren M., Barberıfmmode\\acutea\\elseá\\...         NaN   \n",
       "32                     [Halpin, Kate M., Fishbein, Mark]         NaN   \n",
       "...                                                  ...         ...   \n",
       "12153          [Haufler, Christopher H., Ranker, Tom A.]         NaN   \n",
       "12154                                 [Dubuisson, J.-Y.]         NaN   \n",
       "12163  [Del-Pino, Ivonne Sıfmmode\\acutea\\elseá\\finche...         NaN   \n",
       "12164  [Roquet, Cristina, Sıfmmode\\acutea\\elseá\\fiez,...         NaN   \n",
       "12165  [Bell, Charles D., Calderon, Gloria, Gonzalez,...         NaN   \n",
       "\n",
       "                                  c.doi c.entrytype c.eprint  \\\n",
       "0      10.1111/j.1365-2699.2005.01388.x     article      NaN   \n",
       "23                  10.3732/ajb.1000516     article      NaN   \n",
       "26                   10.7717/peerj.4299     article      NaN   \n",
       "29          10.1016/j.ympev.2021.107110     article      NaN   \n",
       "32             10.1600/036364413X674850     article      NaN   \n",
       "...                                 ...         ...      ...   \n",
       "12153                               NaN     article      NaN   \n",
       "12154            10.1006/mpev.1997.0414     article      NaN   \n",
       "12163        10.1600/036364409787602401     article      NaN   \n",
       "12164        10.1600/036364408783887465     article      NaN   \n",
       "12165          10.1600/036364415X686611     article      NaN   \n",
       "\n",
       "                   c.id c.isbn  ... c.note c.number     c.pages  \\\n",
       "0          Pirie2006Jan    NaN  ...    NaN        1      31--46   \n",
       "23          Tepe2011Aug    NaN  ...    NaN        8  1356--1365   \n",
       "26       Saarela2018Feb    NaN  ...    NaN      NaN       e4299   \n",
       "29         Orton2021Jun    NaN  ...    NaN      NaN      107110   \n",
       "32        Halpin2013Dec    NaN  ...    NaN        4   996--1011   \n",
       "...                 ...    ...  ...    ...      ...         ...   \n",
       "12153    Haufler1995Oct    NaN  ...    NaN        4    361--374   \n",
       "12154  Dubuisson1997Oct    NaN  ...    NaN        2    128--138   \n",
       "12163   Del-Pino2009Mar    NaN  ...    NaN        1      57--67   \n",
       "12164     Roquet2008Mar    NaN  ...    NaN        1    203--217   \n",
       "12165       Bell2015Feb    NaN  ...    NaN        1    327--335   \n",
       "\n",
       "                                 c.publisher  \\\n",
       "0                     John Wiley & Sons, Ltd   \n",
       "23                    John Wiley & Sons, Ltd   \n",
       "26                                PeerJ Inc.   \n",
       "29                            Academic Press   \n",
       "32     American Society of Plant Taxonomists   \n",
       "...                                      ...   \n",
       "12153                  American Fern Society   \n",
       "12154                         Academic Press   \n",
       "12163  American Society of Plant Taxonomists   \n",
       "12164  American Society of Plant Taxonomists   \n",
       "12165  American Society of Plant Taxonomists   \n",
       "\n",
       "                                                 c.title  \\\n",
       "0      `Andean-centred′ genera in the short-branch cl...   \n",
       "23     A 10-gene phylogeny of Solanum section Herpyst...   \n",
       "26     A 250 plastome phylogeny of the grass family (...   \n",
       "29     A 313 plastome phylogenomic analysis of Pooide...   \n",
       "32     A Chloroplast Phylogeny of Agavaceae subfamily...   \n",
       "...                                                  ...   \n",
       "12153  RbcL Sequences Provide Phylogenetic Insights a...   \n",
       "12154  rbcL Sequences: A Promising Tool for the Molec...   \n",
       "12163  trnL-F and rpl16 Sequence Data and Dense Taxon...   \n",
       "12164  Natural Delineation, Molecular Phylogeny and F...   \n",
       "12165  Resolving Relationships within Valerianaceae (...   \n",
       "\n",
       "                                     c.url c.volume c.year s.n_authors  \\\n",
       "0                                      NaN       33   2006           5   \n",
       "23                                     NaN       98   2011           3   \n",
       "26                                     NaN        6   2018          10   \n",
       "29                                     NaN      159   2021           7   \n",
       "32                                     NaN       38   2013           2   \n",
       "...                                    ...      ...    ...         ...   \n",
       "12153  http://www.jstor.org/stable/1547815       85   1995           2   \n",
       "12154                                  NaN        8   1997           1   \n",
       "12163                                  NaN       34   2009           3   \n",
       "12164                                  NaN       33   2008           6   \n",
       "12165                                  NaN       40   2015           5   \n",
       "\n",
       "      v.simil_titles  \n",
       "0              100.0  \n",
       "23             100.0  \n",
       "26             100.0  \n",
       "29             100.0  \n",
       "32             100.0  \n",
       "...              ...  \n",
       "12153          100.0  \n",
       "12154           89.0  \n",
       "12163          100.0  \n",
       "12164           91.0  \n",
       "12165           93.0  \n",
       "\n",
       "[3795 rows x 23 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export database as csv\n",
    "final_df.to_csv(\"bare_final_database.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Definition function to search tematics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function definition for tematic addition\n",
    "def add_tematic_column(file, regex, fields_to_search, search_title, overwrite=True, sep=\",\", index_col=0):\n",
    "    \"\"\"\n",
    "    Open database and search reges in fields_to_search and append column\n",
    "    \"\"\"\n",
    "    # load database\n",
    "    df = pd.read_csv(file, low_memory = True, delimiter=sep, index_col=index_col)\n",
    "\n",
    "    # code title\n",
    "    search_title = f\"t.{search_title}\"\n",
    "    search_title= re.sub(\"\\s\", \"_\", search_title)\n",
    "\n",
    "    # create empty column for result\n",
    "    df[search_title] = np.nan\n",
    "\n",
    "    # iterate row by row\n",
    "    for index, row in df.iterrows(): \n",
    "        # search every pattern in title\n",
    "        for field in fields_to_search:\n",
    "            try: \n",
    "                if re.search(regex, row[field], re.IGNORECASE):\n",
    "                    #add Auto in the cell to indicate this part of the search is true\n",
    "                    df.loc[index, search_title] = \"1\"\n",
    "            except:\n",
    "                continue\n",
    "            \n",
    "    if overwrite:\n",
    "        df.to_csv(file)\n",
    "        return f\"Found {df[df[search_title].notnull()][search_title].shape[0]} for {search_title} regex ({regex}) in {fields_to_search}, {file} overwrote!\"\n",
    "    else:\n",
    "        return df[df[search_title].notnull()][search_title]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Found 252 for t.Mexico regex ((?<!new)(?<!nuevo)\\\\s+m.[xj]ic) in ['c.abstract', 'c.title'], bare_final_database.csv overwrote!\""
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#test function with Mexico regex\n",
    "add_tematic_column(\"bare_final_database.csv\", categories[list(categories.keys())[0]], [\"c.abstract\", \"c.title\"], list(categories.keys())[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add tematics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate over ALL categories. Replace previous results!\r\n",
    "for category in categories:\r\n",
    "    add_tematic_column(\"bare_final_database.csv\", categories[category], [\"c.abstract\", \"c.title\"], category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For individual tematics use add_tematic_column method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual families extraction -- Should be replaced by another method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get families from abstracts\n",
    "\n",
    "file = \"bare_final_database.csv\"\n",
    "df = pd.read_csv(file, low_memory = True, index_col=0)\n",
    "\n",
    "fields_to_search = [\"c.abstract\", \"c.title\"]\n",
    "\n",
    "#Load list of families by major groups\n",
    "with  open(\"../Data_ThePlantList/Angiosperms.txt\", \"r\") as txtfile: \n",
    "    angiosperms = [line.strip() for line in txtfile]\n",
    "    \n",
    "with  open(\"../Data_ThePlantList/Bryophytes.txt\", \"r\") as txtfile: \n",
    "    bryophytes = [line.strip() for line in txtfile]\n",
    "    \n",
    "with  open(\"../Data_ThePlantList/Gymnosperms.txt\", \"r\") as txtfile: \n",
    "    gymnosperms = [line.strip() for line in txtfile]\n",
    "    \n",
    "with  open(\"../Data_ThePlantList/Pteridophytes.txt\", \"r\") as txtfile: \n",
    "    pteridophytes = [line.strip() for line in txtfile]\n",
    "\n",
    "all_families = angiosperms + bryophytes + gymnosperms + pteridophytes\n",
    "\n",
    "for index, row in df.iterrows(): \n",
    "    families = []\n",
    "    groups = []\n",
    "    for field in fields_to_search:\n",
    "        try:\n",
    "#             [families.append(i.lower()) for i in re.findall(\"[A-Za-z]+(?<=ceae)\", row[field], re.IGNORECASE) if i not in families]\n",
    "            for family in re.findall(\"[A-Za-z]+(?<=ceae)\", row[field], re.IGNORECASE):\n",
    "                family_to_add = \"\"\n",
    "                #check family and correct\n",
    "                family_corrected = [s for s in all_families if fuzz.token_sort_ratio(s, family) >= 80]\n",
    "                if len(family_corrected) == 1:\n",
    "                    family_to_add = family_corrected[0]\n",
    "#                     print(family, \"corrected\", family_corrected[0])\n",
    "                else:                    \n",
    "                    family_to_add = family\n",
    "#                     print(family, \"no corrected\")\n",
    "                \n",
    "                if family_to_add not in families:\n",
    "                    families.append(family_to_add.lower())\n",
    "                \n",
    "                \n",
    "                #check group\n",
    "                if family_to_add in angiosperms:  df.loc[index, \"t.angiosperms\"] = 1\n",
    "                elif family_to_add in bryophytes: df.loc[index, \"t.bryophytes\"] = 1\n",
    "                elif family_to_add in gymnosperms: df.loc[index, \"t.gymnosperms\"] = 1\n",
    "                elif family_to_add in pteridophytes: df.loc[index, \"t.pteridophytes\"] = 1\n",
    "                \n",
    "\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    #add to df\n",
    "    list_of_families = \",\".join(set(families))\n",
    "    df.loc[index, \"t.families\"] = list_of_families\n",
    "\n",
    "df.to_csv(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add type of journal (open source, scielo), country, Quartil and Impact factor (year based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search journal in SCIMago databases and add info to the main database\n",
    "\n",
    "file = \"bare_final_database.csv\"\n",
    "df = pd.read_csv(file, low_memory = True, index_col=0)\n",
    "\n",
    "#iterarte by year to open only once each big database\n",
    "for year in range(1999, 2021):\n",
    "    \n",
    "    print(year)\n",
    "        \n",
    "    scimagoFULL_df = pd.read_csv(f\"../Data_scimago/all_journals/scimagojr {year}.csv\", low_memory = True, delimiter=\";\", index_col=0, error_bad_lines=False)\n",
    "    scimagoOPEN_df = pd.read_csv(f\"../Data_scimago/opensource_journals/scimagojr {year}.csv\", low_memory = True, delimiter=\";\", index_col=0, error_bad_lines=False)\n",
    "    scimagoSCIELO_df = pd.read_csv(f\"../Data_scimago/scielo_journals/scimagojr {year}.csv\", low_memory = True, delimiter=\";\", index_col=0, error_bad_lines=False)\n",
    "\n",
    "    \n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        journal = row[\"c.journal\"]\n",
    "        paper_year = row[\"c.year\"]\n",
    "\n",
    "        #avoid nulls\n",
    "        if isinstance(journal, str):\n",
    "\n",
    "            #check quartil, impact factor and country from general database\n",
    "            result = scimagoFULL_df[scimagoFULL_df[\"Title\"].str.lower() == journal.lower()]\n",
    "\n",
    "            if result.shape[0] > 0:\n",
    "                # check year for information that change yearly\n",
    "                if paper_year == year:\n",
    "                    if not isinstance(result[\"SJR\"].values[0], float):\n",
    "                        _sjr = float(result[\"SJR\"].values[0].replace(',', '.'))\n",
    "                    else:\n",
    "                        _sjr = result[\"SJR\"].values[0]\n",
    "                    df.loc[index, \"s.sjr\"] = _sjr\n",
    "                    df.loc[index, \"s.q\"] = result[\"SJR Best Quartile\"].values[0]\n",
    "                    df.loc[index, \"s.h\"] =  result[\"H index\"].values[0]\n",
    "                    \n",
    "                #other static information, just added\n",
    "                df.loc[index, \"s.country\"] =  result[\"Country\"].values[0]\n",
    "                df.loc[index, \"s.scimago\"] = 1\n",
    "\n",
    "            #check if exist in opensource databases (THIS COULD BE SENSITIVE TO YEAR TOO, CHECK IF CHANGES)\n",
    "            result = scimagoOPEN_df[scimagoOPEN_df[\"Title\"].str.lower() == journal.lower()]\n",
    "            if result.shape[0] > 0:\n",
    "                df.loc[index, \"s.opensource\"] = 1\n",
    "\n",
    "                \n",
    "            #check if exist in scielo database\n",
    "            result = scimagoSCIELO_df[scimagoSCIELO_df[\"Title\"].str.lower() == journal.lower()]\n",
    "            if result.shape[0] > 0:\n",
    "                df.loc[index, \"s.scielo\"] = 1\n",
    "\n",
    "\n",
    "    \n",
    "    df.to_csv(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add markers used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load database\n",
    "file = \"bare_final_database.csv\"\n",
    "df = pd.read_csv(file, low_memory = True, index_col=0)\n",
    "\n",
    "#define where to search\n",
    "fields_to_search = [\"c.abstract\", \"c.title\"]\n",
    "\n",
    "#define some common markers\n",
    "markers = {\n",
    "    \"18S-26S\": [\"18S-26S\"],\n",
    "    \"18S\": [\"[ (]18S\"],\n",
    "    \"ADH2F-ADH3R\": [\"ADH2F-ADH3R\"],\n",
    "    \"AFLPs\": [\"AFLP\", \"Amplified fragment length polymorphism\"],\n",
    "    \"aptF-aptH\": [\"aptF-aptH\"],\n",
    "    \"atpB-rbcL\": [\"atpB-rbcL\"],\n",
    "    \"atpB\": [\"[ (]atpB\"],\n",
    "    \"clpp\": [\"clpp\"],\n",
    "    \"COSII\": [\"COSII\"],\n",
    "    \"ETS\": [\"ETS\"],\n",
    "    \"GBSSI\": [\"GBSSI\"],\n",
    "    \"ITS\": [\"ITS\", \"ITS1-5.8s-ITS2\", \"ITS1\", \"5.8s\", \"ITS2\", \"Internal transcribed spacer\", \"59S\"],\n",
    "    \"matK\": [\"matK\"],\n",
    "    \"ndhA\": [\"ndhA\"],\n",
    "    \"ndhF\": [\"ndhF\"],\n",
    "    \"petB\": [\"petB\"],\n",
    "    \"pgiC\": [\"pgiC\"],\n",
    "    \"psaA-ycf3\": [\"psaA-ycf3\"],\n",
    "    \"psaC-ndhE\": [\"psaC-ndhE\"],\n",
    "    \"psaJ-rpl33\": [\"psaJ-rpl33\"],\n",
    "    \"psbE-petL\": [\"psbE-petL\"],\n",
    "    \"psbJ‐petA\": [\"psbJ‐petA\"],\n",
    "    \"psbL-trnS\": [\"psbL-trnS[-AUGC]*\"],\n",
    "    \"psbM-trnD\": [\"psbM-trnD[-AUGC]*\"],\n",
    "    \"psbZ-trnG\": [\"psbZ-trnG[-AUGC]*\"],\n",
    "    \"rbcL\": [\"[ (]rbcL\",\"RuBisCO\", \"rubisco\"],\n",
    "    \"rpl12-clpp\": [\"rpl12-clpp\"],\n",
    "    \"rpl32-trnL\": [\"rpl32-trnL[-AUGC]*\"],\n",
    "    \"rpoB-psbZ\": [\"rpoB[-AUGC]*-psbZ[-AUGC]*\", \"BZ\"],\n",
    "    \"rpoB-trnC-GCA\": [\"rpoB-trnC[-AUGC]*\"],\n",
    "    \"rpoC2-rpoC1\": [\"rpoC2-rpoC1\"],\n",
    "    \"rps16-trnQ\": [\"rps16-trnQ[-AUGC]*\"],\n",
    "    \"rps19\": [\"rps19\"],\n",
    "    \"rps2-rpoc2\": [\"rps2-rpoc2\"],\n",
    "    \"rps4-trnS \": [\"rps4-trnS[-AUGC]*\"],\n",
    "    \"trnC-ycf6\": [\"trnC[-AUGC]*-ycf6\"],\n",
    "    \"trnD-psbM\": [\"trnD[-AUGC]*-psbM\"],\n",
    "    \"trnD-trnT\": [\"trnD[-AUGC]*‐trnY[-AUGC]*‐trnE[-AUGC]*‐trnT[-AUGC]*\"],\n",
    "    \"trnD-trnY\": [\"trnD-AUGC]*-trnY-AUGC]*\"],\n",
    "    \"trnD‐trnY\": [\"trnD[-AUGC]*‐trnY[-AUGC]*\"],\n",
    "    \"trnG-trnfM\" : [\"trnG[-AUGC]*-trnfM[-AUGC]*\"],\n",
    "    \"trnH-psbA\": [\"trnH[-AUGC]*-psbA\", \"psbA-trnH[-AUGC]*\"],\n",
    "    \"trnK-rps16\": [\"trnK[-AUGC]*-rps16\"],\n",
    "    \"trnL-trnF\": [\"trnL[-AUGC]*-trnF[-AUGC]*\", \"trnL-F\", \"trnF-trnL\"],\n",
    "    \"trnM-atpE\": [\"trnM[-AUGC]*-atpE\"],\n",
    "    \"trnP-psaJ\": [\"trnP[-AUGC]*-psaJ\"],\n",
    "    \"trnQ-psbK\": [\"trnQ[-AUGC]*-psbK[-AUGC]*\"],\n",
    "    \"trnR-atpA\": [\"trnR[-AUGC]*-atpA\"],\n",
    "    \"trnS-psbZ\": [\"trnS[-AUGC]*-psbZ[-AUGC]*\"],\n",
    "    \"trnS-rps4\": [\"trnS[-AUGC]*-rps4\"],\n",
    "    \"trnS-trnG\": [\"trnS[-AUGC]*-trnG[-AUGC]*\"],\n",
    "    \"trnS‐trnfM \": [\"trnS[-AUGC]*‐trnfM[-AUGC]*\"],\n",
    "    \"trnT-trnE\": [\"trnT[-AUGC]*-trnE[-AUGC]*\", \"trnE[-AUGC]*‐trnT[-AUGC]*\"],\n",
    "    \"trnT-trnF\": [\"trnT[-AUGC]*-trnF[-AUGC]*\", \"trnF[-AUGC]*‐trnT[-AUGC]*\"],\n",
    "    \"trnT-trnL\": [\"trnT[-AUGC]*-trnL[-AUGC]*\", \"trnT-L\"],\n",
    "    \"trnT-trntL\": [\"trnT[-AUGC]*-trntL[-AUGC]*\"],\n",
    "    \"trnY‐trnE\": [\"trnY[-AUGC]*‐trnE[-AUGC]*\"],\n",
    "    \"ycf1\": [\"ycf1\"],\n",
    "    \"ycf15-ycf1\": [\"ycf15-ycf1\"],\n",
    "    \"ycf4-cemA\": [\"ycf4-cemA\"],\n",
    "    \"Ypr10\": [\"Ypr10\"],\n",
    "    \"Microsatellites\": [\"Microsat.ll?ite\"],\n",
    "          }\n",
    "\n",
    "#iterate row by row\n",
    "for index, row in df.iterrows(): \n",
    "    #create empty result for each row\n",
    "    result_markers = []\n",
    "    #iterate for all markers in dictionary\n",
    "    for marker in markers.keys():\n",
    "        \n",
    "        #search in each field defined\n",
    "        for field in fields_to_search:\n",
    "            #replace all ndash and mdash to normal dash en each field\n",
    "            string_field = row[field]\n",
    "            for r in ((\"–\", \"-\"), (\"—\", \"-\")):\n",
    "                if isinstance(string_field, str):\n",
    "                    string_field = string_field.replace(*r)\n",
    "                \n",
    "            #join string_fieldin all synonyms for a given marker\n",
    "            regex_string = \"|\".join(markers[marker])\n",
    "            #search for all elements\n",
    "            try:\n",
    "                search = re.findall(regex_string, string_field)\n",
    "\n",
    "                if search:\n",
    "                    result_markers.append(marker)\n",
    "            except:\n",
    "                continue\n",
    "\n",
    "    if result_markers:\n",
    "        df.loc[index, \"t.markers\"] = \",\".join(set(result_markers))\n",
    "    else:\n",
    "        df.loc[index, \"t.markers\"] = np.nan\n",
    "\n",
    "                \n",
    "df.to_csv(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add species per study and check mexican endmics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using a.oldtitle search in the original databases in 1 and 1b.\n",
    "# Iterate each row in bare_final_database against every species in both 1 and 1b files\n",
    "# I item in a.oldtitle is found in Paper titles, add \"Species after GB analysis\" t.sp_gb to a columns and add \"Searched name\" to an independent column (t.species) (the latter is the Villaseñor species name)\n",
    "# additionally check with Entrez to get family and other info about names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load database\n",
    "file = \"bare_final_database.csv\"\n",
    "df = pd.read_csv(file, low_memory = True, index_col=0)\n",
    "\n",
    "#put endemic species into a list\n",
    "with open(\"../Data_endemicSpeciesMexico/endemics_uniques.txt\", \"r\") as f:\n",
    "    data = f.read()\n",
    "    endemic_mx = data.splitlines()\n",
    "\n",
    "list_endemic_species = []\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    result_species = []\n",
    "    result_species_endemic = []\n",
    "\n",
    "    #open raw database file\n",
    "    with open(\"../1_Datataxa_extraction/FIRSTTEST.csv\", mode=\"r\") as file:\n",
    "        #load lines\n",
    "        lines = file.readlines()\n",
    "        #iterate over each row (aka species)\n",
    "        for line in lines[1:]:\n",
    "            publications = None\n",
    "            species = None\n",
    "            #manually parse each column\n",
    "            cells = line.split(\"\\\",\\\"\")\n",
    "            if len(cells) >= 2:\n",
    "                publications = cells[2].split(\"|\")\n",
    "            if len(cells) >= 5:\n",
    "                species = cells[5][:-2]\n",
    "            \n",
    "            if publications:\n",
    "                #Check if old title is in publications in raw database\n",
    "                if row[\"a.oldtitle\"] in publications:\n",
    "                    result_species.append(species)\n",
    "                    if species in endemic_mx:\n",
    "                        result_species_endemic.append(species)\n",
    "                        list_endemic_species.append(species)\n",
    "                        df.loc[index, \"s.includedEndemicMXSpp\"] = 1\n",
    "    \n",
    "    df.loc[index, \"t.species\"] = \",\".join(result_species)\n",
    "    df.loc[index, \"t.endemicMXspecies\"] = \",\".join(result_species_endemic)\n",
    "\n",
    "df.to_csv(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"bare_final_database.csv\"\n",
    "df = pd.read_csv(file, low_memory = True, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"t.species\",\"t.endemicMXspecies\", \"s.includedEndemicMXSpp\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "dd67be7cde95a82f2c6722b61600a7fc228d2c9cced8586fb06f345deb16a218"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "metadata": {
   "interpreter": {
    "hash": "dd67be7cde95a82f2c6722b61600a7fc228d2c9cced8586fb06f345deb16a218"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
